url933dd5727b_AppleLiang_Testing_Project_tgz-pJ8-org_joda_time_chrono_gj_MainTestJ8/src/org/joda/time/tz/TestCompiler.java

```diff
@@ -0,0 +1,224 @@
+        "            -8:00   US  P%sT    1946\n" + 
+        "            -8:00   CA  P%sT    1967\n" + 
+        "            -8:00   US  P%sT";
+
+    private DateTimeZone originalDateTimeZone = null;
+
+    public TestCompiler(String name) {
+        super(name);
+    }
+
+    protected void setUp() throws Exception {
+        originalDateTimeZone = DateTimeZone.getDefault();
+        DateTimeZone.setDefault(DateTimeZone.UTC);
+    }
+
+    protected void tearDown() throws Exception {
+        DateTimeZone.setDefault(originalDateTimeZone);
+    }
+
+    //-----------------------------------------------------------------------
+    public void testDateTimeZoneBuilder() throws Exception {
+        // test multithreading, issue #18
+        getTestDataTimeZoneBuilder().toDateTimeZone("TestDTZ1", true);
+        final DateTimeZone[] zone = new DateTimeZone[1];
+        Thread t = new Thread(new Runnable() {
+            public void run() {
+                zone[0] = getTestDataTimeZoneBuilder().toDateTimeZone("TestDTZ2", true);
+            }
+        });
+        t.start();
+        t.join();
+        assertNotNull(zone[0]);
+    }
+
+    private DateTimeZoneBuilder getTestDataTimeZoneBuilder() {
+         return new DateTimeZoneBuilder()
+             .addCutover(1601, 'w', 1, 1, 1, false, 7200000)
+             .setStandardOffset(3600000)
+             .addRecurringSavings("", 3600000, 1601, Integer.MAX_VALUE, 'w', 3, -1, 1, false, 7200000)
+             .addRecurringSavings("", 0, 1601, Integer.MAX_VALUE, 'w', 10, -1, 1, false, 10800000);
+    }    
+
+    //-----------------------------------------------------------------------
+    public void testCompile() throws Exception {
+        Provider provider = compileAndLoad(AMERICA_LOS_ANGELES_FILE);
+        DateTimeZone tz = provider.getZone("America/Los_Angeles");
+
+        assertEquals("America/Los_Angeles", tz.getID());
+        assertEquals(false, tz.isFixed());
+        TestBuilder.testForwardTransitions(tz, TestBuilder.AMERICA_LOS_ANGELES_DATA);
+        TestBuilder.testReverseTransitions(tz, TestBuilder.AMERICA_LOS_ANGELES_DATA);
+    }
+
+    private Provider compileAndLoad(String data) throws Exception {
+        File tempDir = createDataFile(data);
+        File destDir = makeTempDir();
+
+        ZoneInfoCompiler.main(new String[] {
+            "-src", tempDir.getAbsolutePath(),
+            "-dst", destDir.getAbsolutePath(),
+            "tzdata"
+        });
+
+        // Mark all files to be deleted on exit.
+        deleteOnExit(destDir);
+
+        return new ZoneInfoProvider(destDir);
+    }
+
+    private File createDataFile(String data) throws IOException {
+        File tempDir = makeTempDir();
+
+        File tempFile = new File(tempDir, "tzdata");
+        tempFile.deleteOnExit();
+
+        InputStream in = new ByteArrayInputStream(data.getBytes("UTF-8"));
+
+        FileOutputStream out = new FileOutputStream(tempFile);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at line 149. The resource is closed at line 151. There are other execution paths that do not contain closure statements, for example, when InputStream.read throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).